mllib/src/main/scala/org/apache/spark/mllib/util/MLUtils.scala

```diff
@@ -554,4 +554,46 @@ object MLUtils extends Logging {
       math.log1p(math.exp(x))
     }
   }
+
+  /**
+   * Sequentially group input elements to groups, and do aggregation within each group.
+   * A group only contains single key, and be of size no greater than the corresponding size.
+   * For example, input keys = [1, 1, 1, 2, 2, 2, 3, 3, 1],
+   * group sizes are: 1->2, 2->5, 3->1,
+   * then the groups are {1, 1}, {1}, {2, 2, 2}, {3}, {3}, {1}.
+   *
+   * @param input input iterator containing (key, value), usually sorted by key
+   * @param getSize group size of each key.
+   * @return aggregated iterator
+   */
+  private[spark] def combineWithinGroups[K, V, U](
```

Comment:
This method should be helpful when implementing algorithms which needs down-sampling, it is similar to down-sampling in AUC (only one column) while it can be used in multi-column cases with variant group sizes.